### PR TITLE
Refactor SetMethod(...)

### PR DESCRIPTION
**Pull request by agnat**
As discussed in #564 here is a first stab at refactoring `SetMethod(...)`. @kkoopa, @bnoordhuis PTAL and let me know what you think.

### CODE REVIEW

**Comment by kkoopa** [comment]
Clever use of the dummy argument to distinguish types while preserving normal type matching. This seems like an elegant solution.

File: nan.h [hunk 1]
```diff
@@ -1852,75 +1852,47 @@ NAN_INLINE void SetInstanceTemplate(
...
   HandleScope scope;
-  v8::Local<v8::FunctionTemplate> t = New<v8::FunctionTemplate>(callback);
+  v8::Local<v8::FunctionTemplate> tpl = New<v8::FunctionTemplate>(callback);
   v8::Local<v8::String> fn_name = New(name).ToLocalChecked();
-  t->SetClassName(fn_name);
-  templ->Set(fn_name, t);
+  tpl->SetClassName(fn_name);
+  // Note(@agnat): Pass an empty T* as discriminator. See note on
+  // SetMethodAux(...) above
+  imp::SetMethodAux(recv, fn_name, tpl, (T*)0);
```
Comment (bnoordhuis) [inline nan.h]: Can you use `static_cast<T*>(0)`?

File: nan.h [hunk 2]
```diff
@@ -1852,75 +1852,47 @@ NAN_INLINE void SetInstanceTemplate(
...
 
+template <typename T, template <typename> class HandleType>
 NAN_INLINE void SetMethod(
-    v8::Local<v8::ObjectTemplate> templ
+    HandleType<T> recv
   , const char *name
   , FunctionCallback callback) {
   HandleScope scope;
-  v8::Local<v8::FunctionTemplate> t = New<v8::FunctionTemplate>(callback);
+  v8::Local<v8::FunctionTemplate> tpl = New<v8::FunctionTemplate>(callback);
```
Comment (bnoordhuis) [inline nan.h]: Renaming the local makes the diff a little noisier than it needs to be.

**Comment by bnoordhuis** [comment]
It's going to be interesting to see if VS 201[023] accept this.

**Comment by kkoopa** [comment]
Visual Studio 2013 has no problems. https://ci.appveyor.com/project/RodVagg/nan/build/1157

File: nan.h (hunk 2 above)
Comment (agnat) [inline nan.h]: Yeah, the diff is all over the place. Not sure why though. I didn't rename the local. Diff just *thinks* I did. :-/

I'll see what I can do.

File: nan.h (hunk 1 above)
Comment (agnat) [inline nan.h]: Sure. Personally in this case I like `(T*)0` better because it so ... idiomatic. Anyway...

**Comment by kkoopa** [comment]
Looking good. I like it. Much cleaner than unrolling everything. I would think this should work on older VS compilers than that in 2013, but have no easy way of testing it. However, since this whole workaround is only strictly necessary for Node 6 and later, and already Node 4 requires a C++11 compiler, it could even be ifdefed around in case it turns out that it break some old compiler for some old version of Node and someone cared enough about it.

@bnoordhuis You'll want something like this for Node core as well, since `NODE_SET_METHOD` suffers from the same problem. There should be no worry about outdated VS versions there, as Node cannot be built with less than VS 2013.

**Comment by mathiask88** [comment]
@kkoopa You could try to add the `--msvs_version=2010` argument to the appveyor.yml in this PR and start a build for this PR. They got VS 08 10 12 13 15 installed on their systems.

**Comment by agnat** [comment]
I'm confident that it will just work... right down to historic, single digit versions of VS. ;)

I didn't invent it. I've seen it somewhere... in a high-quality code base... probably boost. If it really _does_ act up I'll come up with something (else). TIMTOWTDI.

I removed my beautiful application of a C cast in C++, ironed the diff to remove some wrinkles like @bnoordhuis suggested, squashed the thing and added a proper commit message. PTAL.

**Comment by kkoopa** [comment]
LGTM. Unless someone raises a valid objection, I'll merge this and push out a new version by tomorrow.

**Pull request merged**